### PR TITLE
[stacked] Adds `physicsnemo.mesh.remeshing.partition_cells()` to support GLOBE 3D downsampling.

### PR DESCRIPTION
**Pull request by peterdsharpe**
# PhysicsNeMo Pull Request

## Description

This PR adds a `partition_cells()` function which basically assigns every cell to the nearest point in a point cloud. This is needed for GLOBE 3D's surface subsampling algorithm.

This PR depends on #1509 for correctness, hence why it is stacked.

Demo code to explain in a single picture:

```python
"""Visualize partition_cells on a subdivided cow mesh.

Usage:
    uv run partition_viz.py
"""

import pyvista as pv
import torch

from physicsnemo.mesh.primitives.pyvista_datasets import cow
from physicsnemo.mesh.remeshing import partition_cells

### Load cow and subdivide to get a fine mesh
mesh = cow.load()
print(f"Original cow: {mesh.n_cells:,} cells, {mesh.n_points:,} points")

mesh = mesh.subdivide(levels=2, filter="linear")
print(f"Refined cow:  {mesh.n_cells:,} cells, {mesh.n_points:,} points")

### Partition into 50 clusters
n_seeds = 50
torch.manual_seed(42)
indices = torch.randperm(mesh.n_cells)[:n_seeds]
seeds = mesh.cell_centroids[indices]

result = partition_cells(mesh, seeds)
print(f"Partitioned into {n_seeds} clusters")
print(f"Area conservation error: {abs(result.cluster_areas.sum().item() - mesh.cell_areas.sum().item()):.2e}")

### Plot: color each cell by its cluster assignment
cluster_ids = result.assignments.float()

plotter = mesh.draw(
    cell_scalars=cluster_ids,
    cmap="gist_rainbow",
    show_edges=True,
    alpha_points=0.0,
    backend="pyvista",
)
```

Result:

image

## Checklist

- [x] I am familiar with the [Contributing Guidelines](https://github.com/NVIDIA/physicsnemo/blob/main/CONTRIBUTING.md).
- [x] New or existing tests cover these changes.
- [x] The documentation is up to date with these changes.
- [x] The [CHANGELOG.md](https://github.com/NVIDIA/physicsnemo/blob/main/CHANGELOG.md) is up to date with these changes.
- [x] An [issue](https://github.com/NVIDIA/physicsnemo/issues) is linked to this pull request.
- [x] If I am implementing a new model or modifying any existing model, I have followed the [Models Implementation Coding Standards](https://github.com/NVIDIA/physicsnemo/wiki/Coding-standards-for-models-implementation).

## Review Process

**All PRs are reviewed by the PhysicsNeMo team before merging.**

Depending on which files are changed, GitHub may automatically assign a maintainer for review.

We are also testing AI-based code review tools (e.g., Greptile), which may add automated comments with a confidence score.
This score reflects the AI’s assessment of merge readiness and is **not** a qualitative judgment of your work, nor is
it an indication that the PR will be accepted / rejected.

AI-generated feedback should be reviewed critically for usefulness.
You are not required to respond to every AI comment, but they are intended to help both authors and reviewers.
Please react to Greptile comments with 👍 or 👎 to provide feedback on their accuracy.

### CODE REVIEW

File: physicsnemo/mesh/remeshing/_partition.py [hunk 1]
```diff
@@ -0,0 +1,193 @@
...
+      of curvature, this is an excellent approximation.
+    - Every original cell is assigned to exactly one cluster, so
+      ``cluster_areas.sum() == mesh.cell_areas.sum()`` by construction.
+    - If a cluster receives no cells (possible when seeds outnumber cells or
+      cluster heavily), its area is 0, its normal is the zero vector, and its
+      centroid falls back to the seed position.
+    """
+    n_seeds = len(seeds)
+    device = seeds.device
+    dtype = seeds.dtype
```
[the remark this answers is not in the document]
Comment (peterdsharpe) [reply physicsnemo/mesh/remeshing/_partition.py]: Fixed 5669e0e7177fc0c66ea43cc79b9bee00ae04d5ba

File: physicsnemo/mesh/remeshing/_partition.py [hunk 2]
```diff
@@ -0,0 +1,193 @@
...
+    cell_centroids = mesh.cell_centroids  # (M, D)
+    cell_areas = mesh.cell_areas  # (M,)
+    has_normals = mesh.codimension == 1
+
+    ### Assign each cell to its nearest seed via BVH-accelerated search.
+    # Wrap seeds as a point cloud Mesh (single-vertex cells) so the existing
+    # LBVH + expanding-radius search can be reused.
+    # Complexity: O(N log N) build + O(M log N) query.
+    seed_mesh = Mesh(
+        points=seeds,
```
Comment (laserkelvin) [inline physicsnemo/mesh/remeshing/_partition.py]: Is there a guard to make sure `seeds` and `mesh` have the same `dtype` and `device`?

File: physicsnemo/mesh/remeshing/_partition.py [hunk 3]
```diff
@@ -0,0 +1,193 @@
...
+      smooth surfaces where inter-seed spacing is small relative to the radius
+      of curvature, this is an excellent approximation.
+    - Every original cell is assigned to exactly one cluster, so
+      ``cluster_areas.sum() == mesh.cell_areas.sum()`` by construction.
+    - If a cluster receives no cells (possible when seeds outnumber cells or
+      cluster heavily), its area is 0, its normal is the zero vector, and its
+      centroid falls back to the seed position.
+    """
+    n_seeds = len(seeds)
+    device = seeds.device
```
Comment (laserkelvin) [inline physicsnemo/mesh/remeshing/_partition.py:150]: Probably should either guard or promote device, maybe set `seeds.device` to match `mesh.device`?

File: physicsnemo/mesh/remeshing/_partition.py [hunk 4]
```diff
@@ -0,0 +1,193 @@
...
+    # Complexity: O(N log N) build + O(M log N) query.
+    seed_mesh = Mesh(
+        points=seeds,
+        cells=torch.arange(n_seeds, device=device).unsqueeze(1),
+    )
+    bvh = BVH.from_mesh(seed_mesh)
+    assignments, _ = find_nearest_cells(seed_mesh, cell_centroids, bvh=bvh)
+
+    ### Accumulate areas per cluster
+    cluster_areas = torch.zeros(n_seeds, dtype=cell_areas.dtype, device=device)
```
Comment (laserkelvin) [inline physicsnemo/mesh/remeshing/_partition.py]: Is there a reason why it's `cell_areas.dtype` and not `dtype?

File: physicsnemo/mesh/remeshing/_partition.py [hunk 5]
```diff
@@ -0,0 +1,193 @@
...
+        (exactly the quantity that Lloyd's algorithm drives to zero).
+    """
+
+    assignments: Int[torch.Tensor, " n_cells"]
+    cluster_areas: Float[torch.Tensor, " n_seeds"]
+    cluster_normals: Float[torch.Tensor, "n_seeds n_dims"]
+    cluster_centroids: Float[torch.Tensor, "n_seeds n_dims"]
+
+
+def partition_cells(
```
Comment (laserkelvin) [inline physicsnemo/mesh/remeshing/_partition.py:70]: Do you foresee different partitioning schemes, e.g. geodesics instead of Euclidean distance

File: test/mesh/remeshing/test_partition.py
Comment (laserkelvin) [inline test/mesh/remeshing/test_partition.py:1]: Methinks some GPU tests would be nice in here, maybe, particularly just to make sure we catch scenarios where `seed` and `mesh` are on different devices, different dtypes, etc.

File: physicsnemo/mesh/remeshing/_partition.py (hunk 4 above)
Comment (peterdsharpe) [reply physicsnemo/mesh/remeshing/_partition.py]: No - good catch! These will be the same but `dtype` is more readable

File: test/mesh/remeshing/test_partition.py
Comment (peterdsharpe) [reply test/mesh/remeshing/test_partition.py:1]: Good idea, will add!

File: physicsnemo/mesh/remeshing/_partition.py (hunk 2 above)
Comment (peterdsharpe) [reply physicsnemo/mesh/remeshing/_partition.py]: Good catch! I switched over to an implementation based on `nn.functional.knn` that will handle this, but I've also added my own smoke tests here in `test_partition.py` to check that dtype and device mismatches are handled.

Comment (peterdsharpe) [reply physicsnemo/mesh/remeshing/_partition.py]: Added validation inline as well f291a0237879f002aa0d495300ca5229d9103e89

File: physicsnemo/mesh/remeshing/_partition.py (hunk 4 above)
Comment (peterdsharpe) [reply physicsnemo/mesh/remeshing/_partition.py]: Fixed ce3ef6ed3c1c1aac5150a81d9f5e226cf8b3c87b

File: physicsnemo/mesh/remeshing/_partition.py (hunk 3 above)
Comment (peterdsharpe) [reply physicsnemo/mesh/remeshing/_partition.py:150]: Raising a ValueError here to delegate to the user (in general, swapping devices without telling the user can lead to footguns)

File: test/mesh/remeshing/test_partition.py
Comment (peterdsharpe) [reply test/mesh/remeshing/test_partition.py:1]: 6a47620266763709493b136d3fe47d530f7b7aa3 Added

File: physicsnemo/mesh/remeshing/_partition.py (hunk 5 above)
Comment (peterdsharpe) [reply physicsnemo/mesh/remeshing/_partition.py:70]: Potentially in the future, but geodesics are much slower to calculate in-general, so hoping to defer this to later!
